clients/src/main/java/org/apache/kafka/clients/NetworkClient.java

```diff
@@ -516,13 +516,15 @@ private void initiateConnect(Node node, long now) {
         /* true iff there is a metadata request that has been sent and for which we have not yet received a response */
         private boolean metadataFetchInProgress;
 
-        /* the last timestamp when no broker node is available to connect */
-        private long lastNoNodeAvailableMs;
```

Comment:
This is a slight change in behavior, right? Previously if we cannot find any node to connect to, we'll block in poll() for the retry backoff. Now it seems we'll end up calling `poll()` in a tight loop until we can find a node to connect to. Maybe the logic in `maybeUpdate` needs to be something like this:

1. Check if there is a connection in progress. If so, return the reconnect backoff as the timeout.
2. If there is no connection in progress, call `leastLoadedNode` to find some node to connect to. If this returns null, return retry backoff as is currently done.
3. Next, if we have a node to connect to, then we initiate the connect (if disconnected) and set `lastConnectAttemptNodeId`. 
4. Finally, if already connected, send the metadata request.

Would that keep the current backoff behavior while avoiding the unnecessary connection attempts?